operator/pkg/translate/strategic_port_merge_test.go

```diff
@@ -0,0 +1,141 @@
+		Port:     15021,
+	}
+	istioMetricsPort = &v1.ServicePort{
+		Name:     "metrics-port",
+		Protocol: v1.ProtocolTCP,
+		Port:     15020,
+	}
+)
+
+func TestStrategicPortMergeByPortAndProtocol(t *testing.T) {
+	for _, tt := range []struct {
+		name                string
+		basePorts           []*v1.ServicePort
+		overlayPorts        []*v1.ServicePort
+		expectedMergedPorts []*v1.ServicePort
+	}{
+		{
+			name:                "both base and overlay are nil",
+			basePorts:           nil,
+			overlayPorts:        nil,
+			expectedMergedPorts: nil,
+		},
+		{
+			name:                "overlay is nil",
+			basePorts:           []*v1.ServicePort{httpPort, httpsPort, quicPort},
+			overlayPorts:        nil,
+			expectedMergedPorts: []*v1.ServicePort{httpPort, httpsPort, quicPort},
+		},
+		{
+			name:                "base is nil",
+			basePorts:           nil,
+			overlayPorts:        []*v1.ServicePort{httpPort, httpsPort, quicPort},
+			expectedMergedPorts: []*v1.ServicePort{httpPort, httpsPort, quicPort},
+		},
+		{
+			name:                "same base and overlay",
+			basePorts:           []*v1.ServicePort{httpPort, httpsPort},
+			overlayPorts:        []*v1.ServicePort{httpsPort, httpPort},
+			expectedMergedPorts: []*v1.ServicePort{httpPort, httpsPort},
+		},
+		{
+			name:                "base and overlay for the same port, different protocol",
+			basePorts:           []*v1.ServicePort{httpPort, httpsPort, mysqlPort},
+			overlayPorts:        []*v1.ServicePort{quicPort},
+			expectedMergedPorts: []*v1.ServicePort{httpPort, httpsPort, quicPort, mysqlPort},
+		},
+		{
+			name:                "base and overlay with different ports",
+			basePorts:           []*v1.ServicePort{httpPort},
+			overlayPorts:        []*v1.ServicePort{httpsPort},
+			expectedMergedPorts: []*v1.ServicePort{httpPort, httpsPort},
+		},
+		{
+			name:                "implicit ports",
+			basePorts:           []*v1.ServicePort{httpPort},
+			overlayPorts:        []*v1.ServicePort{httpNoProtoPort},
+			expectedMergedPorts: []*v1.ServicePort{httpPort},
+		},
+		{
+			name:                "status and metrics port are present",
+			basePorts:           []*v1.ServicePort{istioHealthcheckPort, istioMetricsPort, httpsPort},
+			overlayPorts:        []*v1.ServicePort{httpsPort, httpPort},
+			expectedMergedPorts: []*v1.ServicePort{istioHealthcheckPort, istioMetricsPort, httpPort, httpsPort},
+		},
+		{
+			name:                "status port is present",
+			basePorts:           []*v1.ServicePort{istioHealthcheckPort, httpsPort, httpPort},
+			overlayPorts:        []*v1.ServicePort{httpsPort, httpPort},
+			expectedMergedPorts: []*v1.ServicePort{istioHealthcheckPort, httpPort, httpsPort},
+		},
+		{
+			name:                "metrics port is present",
+			basePorts:           []*v1.ServicePort{istioMetricsPort, httpsPort, httpPort},
+			overlayPorts:        []*v1.ServicePort{httpsPort, httpPort},
+			expectedMergedPorts: []*v1.ServicePort{istioMetricsPort, httpPort, httpsPort},
+		},
```

Comment:
need another case here where an existing port number is changed by overlay i.e. ensure we're not breaking existing function.

Reply:
I didn't get it. Here, merging is by port and protocol, not the name. Or did you mean using `overlays` section?

Reply:
I added more cases
* overlay with port name changed (in overlay)
* overlay with different protocol (people don't notice the difference in behavior as long as they are using only TCP for the same protocol. This affects the case when same port, but different protocol is used. It was incorrect before anyways)
* same base and overlay with protocol omitted for overlay (protocol TCP must be present post-merging)
* same base and overlay with protocol omitted for base (protocol TCP must be present post-merging)

Let me know if I have to add more cases.